### PR TITLE
implementing basic skill level service

### LINKED ISSUE #4: Skill level service backend implementation
## Parent

- [x] #3 

## Description
Implement the skill level service in the backend.

## Acceptance Criteria
- [ ] service is implemented

### PR DESCRIPTION
**Pull request by Keyshal-211**
## Description of changes
Implemented the skill level service.

  
## How has this been tested?
Please describe the test strategy you followed.

- [x] automated unit test
- [ ] automated integration test
- [ ] automated acceptance test
- [ ] manual, exploratory test

Test Coverage is low because there are some classes in the project that have not been tested because they are duplicated code and will soon be moved out into a separate repo.
    
## Checklist before requesting a review
- [x] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My code fulfills all acceptance criteria
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [ ] I have made corresponding changes to the documentation
- [x] I have added explanation of architectural decision and rationales to [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/SkillLevelServiceApplication.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.scheduling.annotation.EnableScheduling;
+
+/**
+ * This is the entry point of the application.
+ */
+@SpringBootApplication
+@EnableScheduling
```
Comment (myluki2000) [inline src/main/java/de/unistuttgart/iste/gits/skilllevel_service/SkillLevelServiceApplication.java]: Annotation is unnecessary I think

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/SkillLevelService.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+@RequiredArgsConstructor
+@Transactional
+public class SkillLevelService {
+
+    private final AllSkillLevelsRepository skillLevelsRepository;
+    private final SkillLevelMapper mapper;
+    private final CourseServiceClient courseServiceClient;
+    private final ContentServiceClient contentServiceClient;
+
+    private final RememberLevelCalculator rememberLevelCalculator;
```
Comment (myluki2000) [inline src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/SkillLevelService.java]: You probably do not need separate Calcualtors for the different skill levels because they are all calculated the same way

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/persistence/dao/SkillLevelEntity.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+@NoArgsConstructor
+@AllArgsConstructor
+public class SkillLevelEntity {
+
+    @Id
+    @GeneratedValue
+    private UUID id;
+
+    @Column(nullable = false)
+    private int value;
```
Comment (myluki2000) [inline src/main/java/de/unistuttgart/iste/gits/skilllevel_service/persistence/dao/SkillLevelEntity.java]: The skill level values need to be floats if I remember correctly

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/controller/SkillLevelController.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+@Slf4j
+@Controller
+@RequiredArgsConstructor
+public class SkillLevelController {
+
+    private final SkillLevelService skilllevelService;
+
+    @QueryMapping
+    public SkillLevels userSkillLevels(@Argument UUID courseId, @Argument UUID userId, @Argument UUID chapterId) {
```
Comment (myluki2000) [inline src/main/java/de/unistuttgart/iste/gits/skilllevel_service/controller/SkillLevelController.java]: userSkillLevels query does not have an argument "userId". instead the userId is taken from the graphql context. See the other services how it is done. This means you also need the [RequestHeaderUserInterceptor](https://github.com/IT-REX-Platform/reward_service/blob/main/src/main/java/de/unistuttgart/iste/gits/reward/config/RequestHeaderUserInterceptor.java) class in your config package (you can just take it from the reward service)

File: build.gradle [hunk 5]
```diff
@@ -72,13 +72,21 @@ dependencies {
 			branch = 'main'
 		}
 	}
+	implementation('de.unistuttgart.iste.gits:content_service') { // include content service for DTOs
```
Comment (PaulBredl) [inline build.gradle:75]: would be better to use the tag but whatever

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/SkillLevelServiceApplication.java (hunk 1 above)
Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/skilllevel_service/SkillLevelServiceApplication.java]: Then remove it lol

File: src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            } else {
+                skillLevelToModify.setValue(0);
+            }
+
+        }
+
+        return allSkillLevelsEntity;
+    }
+
+    private SkillLevelEntity getSkillLevelEntityBySkillType(AllSkillLevelsEntity allSkillLevelsEntity,
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/skilllevel_service/service/calculation/SkillLevelCalculator.java:126]: method from hell

**Review by PaulBredl — APPROVED** [review]

**Review by PaulBredl — APPROVED** [review]
